File: HW6/newtonDD.m
% Newton Divided Difference Interpolation

function newtonDD(x, y)
    n = size(x,1);
    n2 = size(x,2);
    if n < n2
        x = x';
        y = y';
        n = n2;
    end
    
    coefs = newtonDDsetup(x, y);
    
    m = 10*n;
    minx = min(x);
    maxx = max(x);
    t = minx:(maxx-minx)/m:maxx;
    
    val = newtonEval(t, coefs, x);
    
    hold off;
    plot(t,val);
    hold on;
    for i=1:1:n
        plot(x(i),y(i),'k*');
    end
end
    
    
% Evaluate divided difference interpolant
function value = newtonEval(t, coefs, x)
    n = size(x,1);
    m = size(t,1);
    for i=1:1:m
        value(i) = coefs(n);
    end
    for i=n-1:-1:1
        value = value .* (t - x(i)) + coefs(i);
    end
end


% Set up divided difference coefficients
function coefs = newtonDDsetup(x, y)
    n = size(x,1);

    % DD level 0
    for i=1:1:n
        coefs(i) = y(i);
    end

    % DD higher levels (bottom to top, overwrite lower entries as they are finished)
    for level=2:1:n
        for i=n:-1:level
            dx = x(i) - x(i-level+1);
            coefs(i) = ( coefs(i)-coefs(i-1) ) / dx;
        end
    end
end

File: HW6/newtonDD.cpp
#include <vector>
using namespace std;

#define TEST

double newtonEval(double t, const vector<double>& coefs, const vector<double>& x) {
  int n = coefs.size();
  double value = coefs[n-1];
  for(int i=n-2; i>=0; i--) {
    value = value * (t - x[i]) + coefs[i];
  }
  return value;
}  

// Set up divided difference coefficients
  int newtonDD(vector<double>& coefs, const vector<double>& x, const vector<double>& y) {
  int n = x.size();
  if(y.size() != n) return 1; // error

  coefs.resize(n);

  // DD level 0
  for(int i=0; i<n; i++) coefs[i] = y[i];

  // DD higher levels (bottom to top, overwrite lower entries as they are finished)
  for(int level=1; level<n; level++) {
    for(int i=n-1; i>=level; i--) {
      double dx = x[i] - x[i-level];
      if(dx==0) return 2;
      
      coefs[i] = ( coefs[i]-coefs[i-1] ) / dx;
    }
  }
  return 0;
}

#ifdef TEST
#define _USE_MATH_DEFINES
#include <iostream>
#include <string>
#include <math.h>

int main() {

    vector<double> x(3);
    x[0] = 0; x[1] = M_PI/4; x[2] = M_PI/2;
    vector<double> y(3);
    y[0] = 1; y[1] = 1/(sqrt(2)); y[2] = 0;
    vector<double> coefs(3);
    
    newtonDD(coefs,x,y);
    
    int n;
    cout << "Enter number of t: ";
    cin >> n;
    double a[n];
    double s[n];
    
    cout << "Enter t's:";
    
    for (int i=0; i<n;i++){
        cin >> a[i];
    }

    for (int j=0; j<n;j++){
        double value = a[j];
        value = fmod(value, 2*M_PI);
        if (value < 0){value += (2*M_PI);}
        
        if (value < (M_PI/2)) {
            //regular value
            s[j] = newtonEval(value,coefs,x);
        } else if (value < M_PI) {
            //pi - value, neg
            value = M_PI - value;
            s[j] = - (newtonEval(value,coefs,x));
        } else if (value < (M_PI + (M_PI/2))) {
            //value - pi, neg
            value = value - M_PI;
            s[j] = - (newtonEval(value,coefs,x));
        } else {
            //2pi - value, pos
            value = (2*M_PI - value);
            s[j] = (newtonEval(value,coefs,x));
        }
        
    }
    cout << "x  " << "cos(x)   " << "cos1(x)   "<< " error" <<"\n"; 
    for (int j=0; j<n;j++){
        
        cout << a[j] << ", " << cos(a[j])
        << ", " << s[j]<< ", " << abs(cos(a[j]) - s[j]) << "\n";
    }
    
  
  return 0;
}
#endif
